File: C++/tests/TestView.cpp
#include <assert.h>
#include <Entity.h>

using namespace std;

struct Vector { float x, y; 
	Vector() {}
	Vector(float m_x, float m_y) : x(m_x), y(m_y) {}
};

EntityCore Core;

Component<Vector>  POSITION 	= Core.newComponent("position", Vector(0.0f, 0.0f));
size_t 			   SPRITE_VIEW 	= Core.newView();
EntityType*        SPRITE  		= Core.newEntityType({POSITION.id}, {}, SPRITE_VIEW);

void testDraw()
{
   	cout << "Running " << __func__ << "() ... ";
   
	string graphics = "OpenGL";

	Entity e(SPRITE);

	e(POSITION).x = 5.1f;
	e(POSITION).y = 2.34f;

	e.draw( &graphics );

	cout << "Pass" << endl;
}

int main()
{
	cout << "Starting " << __FILE__ << "..." << endl;

	Core.setView(SPRITE_VIEW, {POSITION.id}, 
		[] (Entity &e, void *drawState) {
			string graphics = *((string*)drawState);
			Vector p = e(POSITION);
			cout << "Drawing sprite at {" << p.x << ", " << p.y << "} with " << graphics << "." << " ";
		}
	);

	testDraw();

  	cout << __FILE__ << " has ran successfully." << endl;

	return 0;
}

File: Java/test/org/magnos/entity/TestView.java
package org.magnos.entity;

import static org.junit.Assert.assertEquals;

import java.util.concurrent.atomic.AtomicInteger;

import org.junit.Test;
import org.magnos.entity.View.Renderer;
import org.magnos.entity.helper.Vector;

public class TestView
{

	public static final EntityCore core = new EntityCore();
	
	public static final Component<Vector>	POSITION = EntityCore.newComponent( "position", new Vector() );
	public static final View		 		SPRITE_VIEW = EntityCore.newView( "sprite-view" );
	public static final Template			SPRITE = EntityCore.newTemplate( "sprite", new Components(POSITION), new Controllers(), SPRITE_VIEW );
	
	@Test
	public void testDraw()
	{
		final AtomicInteger DRAW_COUNTER = new AtomicInteger();

		EntityCore.setViewDefault( SPRITE_VIEW, new Renderer() {
			public void draw( Entity entity, Object drawState ) {
				DRAW_COUNTER.incrementAndGet();
			}
		});
		
		Entity e = new Entity( SPRITE );
		
		assertEquals( 0, DRAW_COUNTER.get() );
		
		e.draw( null );
		
		assertEquals( 1, DRAW_COUNTER.get() );
	}
	
}
